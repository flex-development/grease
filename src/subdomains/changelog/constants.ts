/**
 * @file Constants
 * @module grease/changelog/constants
 */

import type { ICommitType } from '#src/changelog/interfaces'
import { Type } from '@flex-development/commitlint-config'

/**
 * An array of objects representing explicitly supported commit message types,
 * and whether they should show up in the changelog.
 *
 * @see {@linkcode ICommitType}
 *
 * @const {ICommitType[]} TYPES
 */
export const TYPES: ICommitType[] = [
  { section: ':sparkles: Features', type: Type.FEAT },
  { section: ':bug: Fixes', type: Type.FIX },
  { section: ':mechanical_arm: Refactors', type: Type.REFACTOR },
  { section: ':fire: Performance Improvements', type: Type.PERF },
  { section: ':wastebasket: Reverts', type: Type.REVERT },
  { section: ':pencil: Documentation', type: Type.DOCS },
  { section: ':white_check_mark: Testing', type: Type.TEST },
  { section: ':package: Build', type: Type.BUILD },
  { section: ':robot: Continuous Integration', type: Type.CI },
  { section: ':house_with_garden: Housekeeping', type: Type.CHORE },
  { hidden: true as const, type: Type.RELEASE },
  { hidden: true as const, type: Type.STYLE },
  { hidden: true as const, type: Type.WIP }
]
