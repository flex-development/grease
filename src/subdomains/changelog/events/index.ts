/**
 * @file Entry Point - Events
 * @module grease/changelog/events
 */

export { default as ChangelogEvent } from './changelog.event'
export { default as ChangelogEventListener } from './changelog.listener'
