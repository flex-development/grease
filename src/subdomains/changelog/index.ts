/**
 * @file Entry Point - changelog
 * @module grease/changelog
 */

export { default as ChangelogModule } from './changelog.module'
export * from './constants'
export * from './events'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './queries'
export type * from './types'
