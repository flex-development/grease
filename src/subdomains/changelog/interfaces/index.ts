/**
 * @file Entry Point - Interfaces
 * @module grease/changelog/interfaces
 */

export type { default as IChangelogEntry } from './changelog-entry.interface'
export type { default as ICommitGroup } from './commit-group.interface'
export type { default as ICommitType } from './commit-type.interface'
