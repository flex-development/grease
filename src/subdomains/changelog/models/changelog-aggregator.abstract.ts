/**
 * @file Models - AbstractChangelogAggregator
 * @module grease/changelog/models/AbstractChangelogAggregator
 */

import type { BreakingChange, Commit, Mention, Reference } from '#src/git'
import {
  hasOwn,
  select,
  type CompareResult,
  type PropertyKey
} from '@flex-development/tutils'
import type CommitGroup from './commit-group.model'
import type CommitType from './commit-type.model'

/**
 * Abstract changelog aggregator.
 *
 * @template T - Parsed commit type
 *
 * @abstract
 * @class
 */
abstract class AbstractChangelogAggregator<T extends Commit = Commit> {
  /**
   * Parsed commits.
   *
   * @public
   * @readonly
   * @instance
   * @member {ReadonlyArray<T>} commits
   */
  public readonly commits: readonly T[]

  /**
   * An array of objects representing explicitly supported commit message types,
   * and whether they should show up in the changelog.
   *
   * @see {@linkcode CommitType}
   *
   * @protected
   * @readonly
   * @instance
   * @member {ReadonlyArray<CommitType>} types
   */
  protected readonly types: readonly CommitType[]

  /**
   * Create a new changelog aggregator.
   *
   * @param {ReadonlyArray<CommitType>} types - Commit message types
   * @param {ReadonlyArray<T>?} [commits=[]] - Parsed commits
   */
  constructor(types: readonly CommitType[], commits: readonly T[] = []) {
    this.commits = Object.freeze(commits)
    this.types = Object.freeze(select(types, t => hasOwn(this.map, t.type)))
  }

  /**
   * Get a breaking changes array.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   * @abstract
   *
   * @return {BreakingChange[]} Breaking changes array
   */
  public abstract get breaks(): BreakingChange[]

  /**
   * Get a sorted commit groups array.
   *
   * @see {@linkcode CommitGroup}
   *
   * @public
   * @abstract
   *
   * @return {CommitGroup<T>[]} Sorted commit groups array
   */
  public abstract get groups(): CommitGroup<T>[]

  /**
   * Get a parsed commit map.
   *
   * @protected
   * @abstract
   *
   * @return {Record<string, T[]>} Parsed commit map
   */
  protected abstract get map(): Record<string, T[]>

  /**
   * Get a mentions array.
   *
   * @see {@linkcode Mention}
   *
   * @public
   * @abstract
   *
   * @return {Mention[]} Mentions array
   */
  public abstract get mentions(): Mention[]

  /**
   * Get an array containing issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   * @abstract
   *
   * @return {Reference[]} References array
   */
  public abstract get references(): Reference[]

  /**
   * Get a commit group key.
   *
   * @protected
   * @abstract
   *
   * @param {T} commit - Parsed commit
   * @return {PropertyKey} Commit group key
   */
  protected abstract groupby(commit: T): PropertyKey

  /**
   * Breaking change sorter.
   *
   * @see {@linkcode BreakingChange}
   *
   * @protected
   * @abstract
   *
   * @param {BreakingChange} brk1 - First breaking change to compare
   * @param {BreakingChange} brk2 - Second breaking change to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortbrk(
    brk1: BreakingChange,
    brk2: BreakingChange
  ): CompareResult

  /**
   * Commit sorter.
   *
   * @protected
   * @abstract
   *
   * @param {T} cmt1 - First commit to compare
   * @param {T} cmt2 - Second commit to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortcmt(cmt1: T, cmt2: T): CompareResult

  /**
   * Commit group sorter.
   *
   * @see {@linkcode CommitGroup}
   *
   * @protected
   * @abstract
   *
   * @param {CommitGroup<T>} g1 - First commit group to compare
   * @param {CommitGroup<T>} g2 - Second commit group to compare
   * @return {CompareResult} Comparison result
   */
  protected abstract sortgrp(
    g1: CommitGroup<T>,
    g2: CommitGroup<T>
  ): CompareResult
}

export default AbstractChangelogAggregator
