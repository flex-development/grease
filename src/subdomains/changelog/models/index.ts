/**
 * @file Entry Point - Models
 * @module grease/changelog/models
 */

export {
  default as AbstractChangelogAggregator
} from './changelog-aggregator.abstract'
export { default as ChangelogAggregator } from './changelog-aggregator.model'
export {
  default as ChangelogEntry,
  type ChangelogEntryDTO
} from './changelog-entry.model'
export {
  default as AbstractChangelogFormatter
} from './changelog-formatter.abstract'
export { default as ChangelogFormatter } from './changelog-formatter.model'
export { default as ChangelogInfile } from './changelog-infile.model'
export {
  default as ChangelogStream,
  type ChangelogStreamDTO
} from './changelog-stream.model'
export { default as CommitGroup } from './commit-group.model'
export { default as CommitType } from './commit-type.model'
