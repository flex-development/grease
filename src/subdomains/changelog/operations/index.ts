/**
 * @file Entry Point - Operations
 * @module grease/changelog/operations
 */

export { default as ChangelogOperationHandler } from './changelog.handler'
export {
  default as ChangelogOperation,
  type ChangelogOperationDTO
} from './changelog.operation'
