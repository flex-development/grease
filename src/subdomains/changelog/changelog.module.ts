/**
 * @file ChangelogModule
 * @module grease/changelog/ChangelogModule
 */

import { LoggerService } from '#src/providers'
import { Module } from '@nestjs/common'
import { ChangelogEventListener } from './events'
import { ChangelogOperationHandler } from './operations'
import { ChangelogQueryHandler } from './queries'

/**
 * Changelog module.
 *
 * @class
 */
@Module({
  providers: [
    ChangelogEventListener,
    ChangelogOperationHandler,
    ChangelogQueryHandler,
    LoggerService
  ]
})
class ChangelogModule {}

export default ChangelogModule
