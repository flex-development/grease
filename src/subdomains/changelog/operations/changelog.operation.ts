/**
 * @file Operations - ChangelogOperation
 * @module grease/changelog/operations/ChangelogOperation
 */

import { ChangelogFormatter } from '#src/changelog/models'
import {
  ChangelogQuery,
  type ChangelogQueryParams
} from '#src/changelog/queries'
import { IsConstructor, IsOptional } from '#src/decorators'
import type { Commit } from '#src/git'
import pathe from '@flex-development/pathe'
import {
  defaults,
  fallback,
  isFalse,
  type Assign,
  type Optional
} from '@flex-development/tutils'
import { IsBoolean, IsString } from 'class-validator'

/**
 * Changelog operation payload transfer object.
 *
 * @template T - Parsed commit type
 */
type ChangelogOperationDTO<T extends Commit = Commit> = Assign<
  Partial<ChangelogOperation<T>>,
  Partial<ChangelogQueryParams<T>>
>

/**
 * Changelog operation model.
 *
 * @see {@linkcode ChangelogQuery}
 *
 * @template T - Parsed commit type
 *
 * @class
 * @extends {ChangelogQuery<T>}
 */
class ChangelogOperation<T extends Commit = Commit> extends ChangelogQuery<T> {
  /**
   * Changelog formatter class.
   *
   * @default ChangelogFormatter
   *
   * @public
   * @instance
   * @member {typeof ChangelogFormatter<T>} Formatter
   */
  @IsConstructor()
  public Formatter: typeof ChangelogFormatter<T>

  /**
   * Read `CHANGELOG` from this file.
   *
   * @public
   * @instance
   * @member {Optional<string>} infile
   */
  @IsString()
  @IsOptional()
  public infile?: Optional<string>

  /**
   * Write changelog entries to this file.
   *
   * @public
   * @instance
   * @member {Optional<string>} outfile
   */
  @IsString()
  @IsOptional()
  public outfile?: Optional<string>

  /**
   * Do not write content to {@linkcode outfile} or {@linkcode process.stdout}.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} quiet
   */
  @IsBoolean()
  public quiet: boolean

  /**
   * Read existing entries from and output new entries to {@linkcode infile}.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} samefile
   */
  @IsBoolean()
  public samefile: boolean

  /**
   * Write content to {@linkcode outfile} instead of {@linkcode process.stdout}.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} write
   */
  @IsBoolean()
  public write: boolean

  /**
   * Create a new changelog operation.
   *
   * @see {@linkcode ChangelogOperationDTO}
   *
   * @param {ChangelogOperationDTO<T>?} [payload] - Operation payload
   */
  constructor(payload?: ChangelogOperationDTO<T>) {
    super(payload)

    const {
      Formatter,
      infile,
      outfile,
      quiet,
      samefile,
      write
    } = defaults(fallback(payload, {}), {
      Formatter: ChangelogFormatter,
      infile: undefined,
      outfile: undefined,
      quiet: false,
      samefile: false,
      write: false
    })

    this.Formatter = Formatter
    this.infile = infile
    this.outfile = outfile
    this.quiet = quiet
    this.samefile = fallback(samefile, !!infile && infile === outfile, isFalse)
    this.write = write

    // silence operation
    this.quiet && (this.write = false)

    // changelog regeneration implies samefile
    !this.releases && this.write && (this.samefile = true)

    // override outfile if changelog should be output to infile
    if (this.samefile) {
      this.outfile = this.infile = fallback(this.infile, 'CHANGELOG.md')
    }

    // resolve infile and outfile relative to cwd
    this.infile && (this.infile = pathe.resolve(this.cwd, this.infile))
    this.outfile && (this.outfile = pathe.resolve(this.cwd, this.outfile))
  }
}

export { ChangelogOperation as default, type ChangelogOperationDTO }
