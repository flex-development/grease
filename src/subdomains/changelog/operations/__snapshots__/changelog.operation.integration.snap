// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:changelog/operations/ChangelogOperation > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isConstructor": "Formatter must be a constructor",
    },
    "property": "Formatter",
    "value": [Function Formatter],
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "infile must be a string",
    },
    "property": "infile",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "outfile must be a string",
    },
    "property": "outfile",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "samefile must be a boolean value",
    },
    "property": "samefile",
    "value": 1,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "write must be a boolean value",
    },
    "property": "write",
    "value": 0,
  },
]
`;
